### ISSUE TITLE
Prediction for a single tree

### ISSUE THREAD

**Issue by PhilippPro** [imbs-hl/ranger#351]
Hi Marvin,

is there a possibility to just predict the outcome for a single tree? I have seen, that there is an option to use the first `num.trees`, but I would like to have an option to just predict e.g. the prediction of the 20th tree. 

I could use this for a function that calculates the variable importance for an arbitrary measure of mlr (currently exported in the measures package). My current prototype can be found here, maybe you have some suggestions: [varImpRanger](https://github.com/PhilippPro/varImp/blob/master/varImpRanger/varImpRanger.R)
I currently use all the trees, which will make it very slow for bigger datasets (and it is already slow for smaller datasets). 

The varImp package is currently only working for `cforest` of the party package.

**Comment by mnwright** [comment]
You are using a new permutation for each tree to avoid predicting with the whole forest for several permutations? 

Anyway, I agree that `predict()` should have a selection of single trees. 
Until this is implemented I have a little hack that does the trick: 

````R
swap_trees <- function(rf, i, j) {
  res <- rf
  
  res$forest$child.nodeIDs[[i]] <- rf$forest$child.nodeIDs[[j]]
  res$forest$split.varIDs[[i]] <- rf$forest$split.varIDs[[j]]
  res$forest$split.values[[i]] <- rf$forest$split.values[[j]]
  
  res$forest$child.nodeIDs[[j]] <- rf$forest$child.nodeIDs[[i]]
  res$forest$split.varIDs[[j]] <- rf$forest$split.varIDs[[i]]
  res$forest$split.values[[j]] <- rf$forest$split.values[[i]]
  
  res
}
````
This function will swap two trees in a forest. You can swap your tree of interest to position 1 and predict with `num.trees = 1`. Example:

````R
rf <- ranger(Species ~., iris, num.trees = 5)
rf2 <- swap_trees(rf, 1, 4)
pred_tree4 <- predict(rf2, iris, num.trees = 1)$predictions

# Check
pred_all <- predict(rf, iris, predict.all = TRUE)$predictions
all(as.numeric(pred_tree4) == pred_all[, 4])
````

Does this help?

**Comment by PhilippPro** [comment]
I think following the definition correctly, I have to permute the observations for each tree seperately. 

Thanks I will try this out. Speed is really an issue in my current implementation. I think your VIMP is currently calculated internally, which is of course much faster.

**Comment by PhilippPro** [comment]
Yes, kind of. That would mean permuting each variable for each tree (but only the OOB observations) and giving the prediction. 

Without permuting I think I already can just predict and get the predictions for the OOB observations so this part is not necessary to do it internally. (This is currently not well programmed in the prototype.)

**Comment by PhilippPro** [comment]
I think your swap function does not work well for probability predictions:

```
rf <- ranger(Species ~., iris, num.trees = 5, probability = TRUE)
rf2 <- swap_trees(rf, 1, 4)
pred_tree4 <- predict(rf2, iris, num.trees = 1)$predictions
# Check
pred_all <- predict(rf, iris, predict.all = TRUE)$predictions
pred_tree4 == pred_all[,,4]
```

**Comment by mnwright** [comment]
I didn't swap the `terminal.class.counts` object which is required for probability predictions. I've edited my answer above to include this. Should be fine now.
